### DISCUSSION TITLE
Use GPG private key within GitHub Actions and Setup Java SDK

### DISCUSSION THREAD

**Discussion by bissim** [community/community#27165] [Actions]
<p>Hello,<br>
I’m trying to deploy my project over Nexus by defining a workflow on GitHub; I already managed to deploy over GitHub Packages with not too much effort.<br>
However, <a href="https://central.sonatype.org/pages/requirements.html" rel="nofollow noopener">Nexus requires release artifacts to be signed with a private key</a> in order to be distributed.<br>
I tried to add both my key and its passphrase as secrets to my repository but with no avail: during build process, maven errors with</p>
<pre><code class="lang-auto">[ERROR] Failed to execute goal org.apache.maven.plugins:maven-gpg-plugin:1.6:sign (sign-artifacts) on project ***: Unable to decrypt gpg passphrase: org.sonatype.plexus.components.sec.dispatcher.SecDispatcherException: java.io.FileNotFoundException: /home/runner/.m2/settings-security.xml (No such file or directory) -&gt; [Help 1]
</code></pre>
<p>I’m not sure about non-existing <code>settings-security.xml</code>, since I’m using the <a href="https://github.com/marketplace/actions/setup-java-jdk" rel="nofollow noopener">Setup Java JDK</a> action and there is no mention of such a file in <a href="https://maven.apache.org/plugins/maven-gpg-plugin/usage.html" rel="nofollow noopener">Maven GPG plugin</a> configuration; looks like <a href="https://maven.apache.org/guides/mini/guide-encryption.html" rel="nofollow noopener">this mechanism</a> takes place, however I have no idea how to deal with it in GitHub Actions.</p>
<p>Moreover, since I use two Setup Java SDK actions within my workflow, I have the problem of the clean-up for the second one: since the first one already deletes my private key, the second one fails in doing this and errors the whole workflow.</p>
<p>So, my questions are:</p>
<ul>
<li>How do I properly set artifact signing with Setup Java JDK action?</li>
<li>How do I avoid failing of two subsequent cleanup for the same kind actions?</li>
</ul>
<p>For reference, <a href="https://raw.githubusercontent.com/bissim/FLY-graph/master/.github/workflows/maven-publish.yml" rel="nofollow noopener">here</a> is workflow for my project.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/bissim">@bissim</a>,</p>
<p>About your two questions:</p>
<ol>
<li>
<blockquote>
<p>How do I properly set artifact signing with Setup Java JDK action?</p>
</blockquote>
<p>The ‘<strong>settings-security.xml</strong>’ file contains the master password used to encrypt remote repository credentials. The default location of this file is ‘<strong>~&lt;user_home&gt;/.m2/settings-security.xml</strong>’.</p>
<p>Normally, if this file does not exist, or if it does not contain valid content, the ‘<a href="https://kb.novaordis.com/index.php/Maven_Command_Line_Options#-ep.2C_--encrypt-password" rel="nofollow noopener"><strong>mvn --encrypt-password</strong></a>’ command line will fail with the error like:</p>
<pre><code class="lang-auto">[ERROR] java.io.FileNotFoundException: ~&lt;user_home&gt;/.m2/settings-security.xml (No such file or directory)
</code></pre>
<p>In this situation, you may need to create this file by yourself, then copy and paste the encrypted master password (output of the command line ‘<strong>mvn --encrypt-master-password</strong>’) into the file.</p>
<p>If you do not store the ‘<strong>settings-security.xml</strong>’ file in the directory ‘<strong>~&lt;user_home&gt;/.m2</strong>’, for example in your project repository directory, you can try to use the following command line to refer to this file:</p>
<pre><code class="lang-auto">mvn -Dsettings.security=path/to/settings-security.xml
</code></pre>
</li>
<li>
<blockquote>
<p>How do I avoid failing of two subsequent cleanup for the same kind actions?</p>
</blockquote>
<p>You can try to divide the current job to two jobs in your workflow. One is for “<strong>Publish GitHub Packages Apache Maven</strong>”, another is for “<strong>Publish to Apache Maven Central</strong>”.<br>
In this way, the Setup Java JDK action is running in different jobs on different runner machines, they will not have any effect or conflict on each other.</p>
</li>
</ol>

**Comment by bissim** [comment]
<p>Hello <a class="mention" href="/brightran">@brightran</a>,<br>
Thank you for your answer.</p>
<p>About number 2, it makes sense: I was avoiding to use multiple jobs in order not to duplicate steps because of two different runners; I think there’s no other way to fix that problem.</p>
<p>About number 1, I wonder whether there’s a way to create a working <code>settings-security.xml</code> file remotely: I guess that prepending a step with <code>mvn --encrypt-master-password</code> command won’t solve the problem, also because it prompts for a password. This is really a problem: I cannot interact with runners, unless I make use of some really bash commands.</p>
<p><strong>Edit:</strong> by the way, why isn’t this problem documented into ‘<a href="https://docs.github.com/en/actions/language-and-framework-guides/publishing-java-packages-with-maven" rel="nofollow noopener">Publishing Java projects with Maven</a>’ documentation?</p>
<p><strong>Edit2:</strong> can’t I just disable password encryption mechanism? In the first place, I don’t need such a level of security, since I’m using a disposable runner and passwords aren’t even encrypted in generated <code>settings.xml</code> file.</p>
<p><strong>Edit3:</strong> I managed to get rid of the error about <code>settings-security.xml</code> by passing server username and password and GPG key passphrase as environment variables rather than directly to Setup Java configuration, like shown in Setup Java <a href="https://github.com/actions/setup-java#publishing-using-apache-maven" rel="nofollow noopener">README</a>. The error I’m getting now is</p>
<pre><code class="lang-auto">gpg: signing failed: Inappropriate ioctl for device
</code></pre>
<p>that cannot be fixed by <code>export GPG_TTY=$(tty)</code>.<br>
So, does anyone know how to fix this issue?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/bissim">@bissim</a>,</p>
<p>In the docs “<a href="https://docs.github.com/en/actions/language-and-framework-guides/publishing-java-packages-with-maven" rel="nofollow noopener">Publishing Java packages with Maven</a>”, it seems only mentions using passwrod (personal access token) to authenticate, not mentions using GPG keys. And looks like, all the authentication information will be stored into the automatically generated <em><strong>settings.xml</strong></em>.<br>
Not sure why the <strong>settings-security.xml</strong> will be involved.</p>
<p>I have created an issue ticket (<a href="https://github.com/actions/setup-java/issues/91" rel="nofollow noopener">actions/setup-java#91</a>) to help you report the questions to the appropriate engineering team for further investigation and evaluation.<br>
You can follow this issue ticket and add your comments on it.</p>

**Comment by bissim** [comment, marked as answer]
<p><a class="mention" href="/brightran">@brightran</a>,<br>
Again, thank you for your concern.</p>
<p>As stated in the <a href="https://github.com/actions/setup-java/issues/91" rel="nofollow noopener">issue</a> you kindly opened on <a href="https://github.com/actions/setup-java" rel="nofollow noopener">actions-setup-java</a> repository, thanks to <a class="mention" href="/airquick">@airquick</a> <a href="https://github.com/actions/setup-java/issues/43#issuecomment-674000825" rel="nofollow noopener">intervention on issue 43</a> which I previously commented, the solution for the <code>ioctl</code> issue is to add a configuration section to Maven GPG plugin in project POM:</p>
<pre><code class="lang-auto">            &lt;configuration&gt;
              &lt;!-- Prevent gpg from using pinentry programs --&gt;
              &lt;gpgArguments&gt;
                &lt;arg&gt;--pinentry-mode&lt;/arg&gt;
                &lt;arg&gt;loopback&lt;/arg&gt;
              &lt;/gpgArguments&gt;
            &lt;/configuration&gt;
</code></pre>
<p>Along with <a href="https://github.community/t/use-gpg-private-key-within-github-actions-and-setup-java-sdk/127133/2">first brightran answer</a> to this thread, this closes my request for support.</p>

**Comment by slawekjaranowski** [comment]
<p>You can also use another plugin <a href="https://www.simplify4u.org/sign-maven-plugin/" class="inline-onebox" rel="noopener nofollow ugc">Sign Maven Plugin – Introduction</a> it doesn’t use <code>gpp</code> for signing.</p>
<p><code>sign-maven-plugin</code> support all configuration item as environment variable, so configuration for <code>CI</code> system like <code>GitHub Actions</code> is more easier and simply work without difficult configuration.</p>
<p>It also is operating system independent because needn’t external software.</p>
